### ISSUE TITLE
autologin for team on startup possible?

### ISSUE THREAD

**Issue by ownagedj** [raelgc/scudcloud#435]
Hi,
is it in possible to have an autologin for teams instead of clicking on the `You're already signed in to this team:`?

Ubuntu 16.04 and latest scudcloud

Thanks in advance

**Comment by raelgc** [comment]
It's how it's supposed to work.

**Comment by e1himself** [comment]
Maybe it's better to auto-login if there's just one team linked? 
Clicking that button again and again feels a bit annoying if I always use Slack for a single team.

**Comment by raelgc** [comment]
@e1himself Sorry if I'm not clear. The way you described - just launch ScudCloud and teams are automatically logged - it's the way it's supposed to work.

Can you, please, remove the config file (`~/.config/scudcloud/scudcloud.cfg`) and start register your teams again?

**Comment by e1himself** [comment]
@raelgc Tried that. But it still requires to click team button. 
Let's use test case description to be 100% clear :)

_Expected behavior_:
- Run Scudcloud app first time
- Login screen displayed: team domain, login and password are prompted
- Log in to your team account
- Close Scudcloud app
- Open Scudcloud app again
- It should be _logged in_ to your team account (displaying channels list on left, and chat on right)

_Actual behavior_:
- Run Scudcloud app first time
- Login screen displayed: team domain, login and password are prompted
- Log in to your team account
- Close Scudcloud app
- Open Scudcloud app again
- It shows login screen again, but with "You're already signed in to this team:" section at bottom.
- You click team name button get app to a desired _logged in_ state

**Comment by raelgc** [comment]
Thank you @e1himself for the clear description. Just trying to wonder why it's not working as expected for you. 

Just to double check: is the `Keep me signed in` field checked?

**Comment by e1himself** [comment]
@raelgc Yes it is checked. And was checked before during writing test case.  Bizarre.

**Comment by e1himself** [comment]
Maybe it's related to the fact we're using Ubuntu 16.04? (me and @ownagedj)

**Comment by raelgc** [comment]
The weird fact: I'm using 16.04 too (and default Unity).

**Comment by ownagedj** [comment]
I removed any config for Scudcloud.
Also did a fresh install.
Checked that the `Keep me signed in` field is set.
Same outcome

**Comment by YaroslavHavrylovych** [comment]
+1 Ubuntu 14.04

**Comment by raelgc** [comment]
If you start scudcloud by the command line, can you put the output here? (On 16.04 a warning message about non specified version for some libraries is expected).

**Comment by ownagedj** [comment]
When I start it via terminal I'll get this

```
/usr/lib/python3/dist-packages/scudcloud/notifier.py:3: PyGIWarning: Notify was imported without specifying a version first. Use gi.require_version('Notify', '0.7') before import to ensure that the right version gets loaded.
  from gi.repository import Notify
/usr/lib/python3/dist-packages/scudcloud/scudcloud.py:27: PyGIWarning: Unity was imported without specifying a version first. Use gi.require_version('Unity', '7.0') before import to ensure that the right version gets loaded.
  from gi.repository import Unity, Dbusmenu
SpellCheck: Language = "en_US" 
```

**Comment by eyalzek** [comment]
@raelgc the output the guy above me described can be fixed by implementing this patch: https://github.com/raelgc/scudcloud/pull/421/files
After fixing it, starting scudcloud from command doesn't give any output besides: `SpellCheck: Language = "en_US"`
which is unrelated I presume.. I'm on Xubuntu 16.04 btw.

**Comment by raelgc** [comment]
@eyalzek Thank you for checking this. 

Yes, it's unrelated. It'll remove the warning on 16.04, but it'll break ScudCloud on 12.04.

The solution then, for 12.04, will be then keep it in a separate branch, which means someone to apply all updates to it. Or create a complicated checking based on Ubuntu version.

**Comment by raelgc** [comment]
@eyalzek Just to double check: this make your autologin works?

**Comment by eyalzek** [comment]
@raelgc No it  doesn't, I'm having the same problem. should've been more clear.

Also, I'm pretty sure I checked the `keep me signed in` option, but I can't seem to find it now

**Comment by raelgc** [comment]
So see it again, only if you remove the config file (`~/.config/scudcloud/scudcloud.cfg`) and start again on the same team. If you don't have this folder, please, create it.

**Comment by eyalzek** [comment]
Tried it now, results are the same.

**Comment by eyalzek** [comment]
Where is it specified in the code btw? I can't find anything related to loading a `keep me signed in` setting, or automatically signing in to a team...

**Comment by raelgc** [comment]
@eyalzek In [scudcloud.py#L122](https://github.com/raelgc/scudcloud/blob/master/scudcloud/scudcloud.py#L122) we're saying that we'll use a custom class ([cookiesjar.py](https://github.com/raelgc/scudcloud/blob/master/scudcloud/cookiejar.py)) to handle cookies. This class will save cookies in a file and restore them on startup.

For this reason you don't have to enter your team again.

But, for some reason, this is not saving the autologin for you guys facing this issue.

If we could find the reason... I really wish fix this annoying issue for you :(

**Comment by eyalzek** [comment]
@raelgc if I can help you guys debug this in any way, let me know.

**Comment by raelgc** [comment]
Can someone run a simple test for this?

Edit, using `sudo`, `/usr/lib/python3/dist-packages/scudcloud/cookiejar.py` and change the below line (`16`):

``` python
                data.append(cookie.toRawForm()+ "\n")
```

replacing it by this:

``` python
                data.append(cookie.toRawForm().append("\n"))
```

Of course, close and restart the program.

**Comment by zQQp** [comment]
I tried it and it did not help unfortunately :(

here is what I did:
- close scudcloud (not to tray of course)
- edit the cookiejar file
- opened scudcloud
- clicked button to enter my team
- close scudcloud  (not to tray of course)
- opened scudcloud again

> > still not logged in automatically :/

**Comment by raelgc** [comment]
I'm reading all posts again. Sounds like you're automatically logged in, but every time you open Scudcloud, you're not automatically redirected to `#general`, right? (As you can see a button stating you're already logged in).

**Comment by eyalzek** [comment]
@raelgc yes, that's correct

**Comment by raelgc** [comment]
@eyalzek What is the output if you start scudcloud with `scudcloud --debug=True &` ?

For me this is the output (aside warnings in 16.04):

``` term
SpellCheck: Language = "en_US" 
URL Changed: https://scudcloud.slack.com/messages
URL Changed: https://scudcloud.slack.com/messages/general/
```

**Comment by vlatkoB** [comment]
Same issue on Arch. Output is only
`URL Changed: https://slack.com/signin`

However, before that I see a warning
`
/usr/lib/python3.5/site-packages/scudcloud/notifier.py:3: PyGIWarning: Notify was imported without specifying a version first. Use gi.require_version('Notify', '0.7') before import to ensure that the right version gets loaded.
  from gi.repository import Notify
`
Not sure if it's related.

Edit:
After I click on channel:
`Mouse Button 1`
`Link Clicked: https://workinghaskell.slack.com/`
`URL Changed: https://workinghaskell.slack.com/messages`
`URL Changed: https://workinghaskell.slack.com/messages/general/`

**Comment by raelgc** [comment]
Thank you @vlatkoB! 

Now, we need to check which is the saved value for the domain.

Please, take a look in the config file: `~/.config/scudcloud/scudcloud.cfg`.

It should have 5 sessions: 
1. `Cookies` occupying several lines (or one giant line), then 
2. `Domain`
3. `Menu`
4. `Systray` 
5. `Zoom`
6. `geometry` and
7. `windowState`

Can you tell me what is the value for `2. Domain`?

**Comment by vlatkoB** [comment]
I have only these, ordered as in my cfg:
1. Cookies
4. Systray
3. Menu
6. geometry
7. windowState
2. Domain and 5. Zoom are not present. 
Installed version is 1.30-1.

**Comment by raelgc** [comment]
Nice, thank you @vlatkoB! Finally we found the reason!

`Zoom` not present is ok. `Domain` missing is the reason why it's not automatically loading your teams.

In order to test this, you can add it to this file, at any new line, like this: 

``` term
Domain=https://scudcloud.slack.com/
```

or several:

``` term
Domain=https://scudcloud.slack.com/, https://example.slack.com/, https://foobar.slack.com/
```

**Comment by vlatkoB** [comment]
:-) Do I have to say more?

**Comment by raelgc** [comment]
@vlatkoB your team has a custom icon? I'm suspecting our app is assuming team always have one specific icon, and this is breaking the logic to save logged domains.

**Comment by zQQp** [comment]
mine has a custom icon and it's not saving automatically.
After I added it to the config file, it works like a charm! THANKS ALOT!

**Comment by eyalzek** [comment]
@raelgc that would make sense, on my home PC with one team it worked fine and on my laptop with another team (that has a custom icon) it doesn't.

**Comment by maciex** [comment]
I have the same problem. I need to click the team link to log in. I also checked the latest Dev Tree, but got the same behaviour.

In scudcloud.cfg I have:

```
$ grep ^Domain ~/.config/scudcloud/scudcloud.cfg
Domain=@Invalid()
```

The value the code is trying to save is empty (`self.domain`).

Edit:
When I added the proper value of `Domain=` to the config file, autologin works on every start of the app.

**Comment by raelgc** [comment]
@maciex Thanks for this info, it'll really helpful! Wondering if the `self.domain` got not assigned, and if yes, the reason. When starting the app by the command line, and then closing it, do you see any errors?

**Comment by maciex** [comment]
@raelgc The app closes properly. The app closes properly.

I added few prints to find what going on but it looks like all the variables are empty:

```
$ scudcloud                                                                                                                                                                                 
/usr/lib/python3.5/site-packages/scudcloud/notifier.py:3: PyGIWarning: Notify was imported without specifying a version first. Use gi.require_version('Notify', '0.7') before import to ensure that the right version gets loaded.                                                                                                                                                                                            
  from gi.repository import Notify                                                                                                                                                                             
loaded the Generic plugin                                                                                                                                                                                      
self.identifier: None                                                                                                                                                                                          
self.domains: []                                                                                                                                                                                               
closeEvent(): self.domains: [] 
```

Edit:
I tried to attach the changed scudcloud.py but the .py/.txt/.zip files shows as not supported. So I'm just pasting a diff:

```
$ diff -u /usr/lib/python3.5/site-packages/scudcloud/scudcloud.py /usr/lib/python3.5/site-packages/scudcloud/scudcloud.py_debug 
--- /usr/lib/python3.5/site-packages/scudcloud/scudcloud.py     2016-10-10 12:59:51.000000000 +0200
+++ /usr/lib/python3.5/site-packages/scudcloud/scudcloud.py_debug       2016-11-10 14:18:44.730797469 +0100
@@ -64,6 +64,8 @@
         centralWidget.setLayout(layout)
         self.setCentralWidget(centralWidget)
         self.startURL = Resources.SIGNIN_URL
+
+        print("self.identifier: %s" % self.identifier)
         if self.identifier is not None:
             if isinstance(self.identifier, str):
                 self.domains = self.identifier.split(",")
@@ -72,6 +74,7 @@
             self.startURL = self.normalize(self.domains[0])
         else:
             self.domains = []
+        print("self.domains: %s" % self.domains)
         self.addWrapper(self.startURL)
         self.addMenu()
         self.tray = Systray(self)
@@ -315,6 +318,7 @@
         if len(self.domains) == 0:
             self.domains.append(teams[0]['team_url'])
         team_list = [t['team_url'] for t in teams]
+        print("team_list: %s" % team_list)
         for t in teams:
             for i in range(0, len(self.domains)):
                 self.domains[i] = self.normalize(self.domains[i])
@@ -331,6 +335,9 @@
                                 self.settings.setValue("Domain", self.domains)
         if len(teams) > 1:
             self.leftPane.show()
+        
+        print("teams: %s" % teams)
+        print("team_list2: %s" % team_list)

     def switchTo(self, url):
         exists = False
@@ -386,6 +393,7 @@
         self.forceClose = True

     def closeEvent(self, event):
+        print("closeEvent(): self.domains: %s" % self.domains)
         if not self.forceClose and self.settings.value("Systray") == "True":
             self.hide()
             event.ignore()
```

**Comment by maciex** [comment]
After some time I checked again and looks like it's working... (scudcloud.noarch 1.33-2.fc24).
I also upgraded (to scudcloud.noarch 1.37-1.fc24 ) and retested and it still works.

**Test description:**
1. Make sure the app is closed
2. Remove "Domain" from config file
3. Restart the app
4. Login to/Select your organization
5. Close (also Tray icon) the app
6. Start the app

**Expected result:**
Scudcloud automatically logs in to my organization.

**Actual result:**
As expected (above).
